lib/rust/parser/src/data/tree.rs

```diff
@@ -0,0 +1,214 @@
+//! A map-based tree for use in the macro resolver.
+
+use crate::prelude::*;
+
+
+
+// ============
+// === Tree ===
+// ============
+
+/// The type of branches in the tree.
+pub type Branches<K,V> = HashMap<K,Tree<K,V>>;
+
+/// A tree with a variable number of branches per node.
+#[derive(Derivative)]
+#[derivative(Clone)]
+#[derivative(Debug(bound="K:Debug+Eq+Hash+PartialEq, V:Debug+Eq+PartialEq"))]
+#[derivative(Default(bound="K:Eq+Hash"))]
+#[derivative(PartialEq(bound="K:Eq+Hash, V:PartialEq"))]
+#[derivative(Eq(bound="K:Eq+Hash, V:Eq"))]
+pub struct Tree<K,V> {
+    /// The value at the current node.
+    pub value : Option<V>,
+    /// The branches at the current node.
+    pub branches : Branches<K,V>
+}
+
+impl<K,V> Tree<K,V> {
+    /// Check if `self` is a leaf of the tree.
+    pub fn is_leaf(&self) -> bool {
+        self.branches.is_empty()
+    }
+
+    /// Check if `self` is a non-leaf node in the tree.
+    pub fn is_non_leaf(&self) -> bool {
+        !self.is_leaf()
+    }
+}
+
+impl<K,V> Tree<K,V> where K:Clone+Eq+Hash+PartialEq {
```

Comment:
Create alias for the types, like `trait Bounds = Clone+Eq+Hash+PartialEq;` which can be accessed fro the outside as `tree::Bounds` - that will make external code much nicer sometimes